### PR TITLE
Access indices array for VBuffer in KeyToVector transformer only when resulting vector is sparse.

### LINKED ISSUE #1751: Spittler/consolidator worker exception while consuming data w/ 0.8.0-preview-27128-5
### Issue

- **What did you do?**
Upgraded from 0.7 to 0.8-preview.

- **What happened?**
I got an exception during `estimator.fit()`. Though it also appears doing a `encodedTrainingData.AsDynamic.Preview` where my "encoder" only does encoding (no ML model learnt).

I'm not sure how to further debug this. I probably won't have time for a minimal repro example.

### Source code / logs
Exception:

**IndexOutOfRangeException: Index was outside the bounds of the array.**

```
                public void SetAll(OutPipe[] pipes)
                {
                    if (_ex != null)
                        throw Contracts.Except(_ex, "Splitter/consolidator worker encountered exception while consuming source data");
                    Contracts.Assert(Utils.Size(pipes) == _batchColumns.Length);
[...]
```

Exception trace:
```
   at System.ThrowHelper.ThrowIndexOutOfRangeException() in E:\A\_work\65\s\corefx\src\System.Memory\src\System\ThrowHelper.cs:line 43
   at Microsoft.ML.Transforms.Conversions.KeyToVectorMappingTransformer.Mapper.<>c__DisplayClass11_0.<MakeGetterOne>b__0(VBuffer`1& dst) in E:\A\_work\423\s\src\Microsoft.ML.Data\Transforms\KeyToVector.cs:line 483
   at Microsoft.ML.Runtime.Data.ColumnConcatenatingTransformer.Mapper.BoundColumn.<>c__DisplayClass18_0`1.<MakeGetter>b__0(VBuffer`1& dst) in E:\A\_work\423\s\src\Microsoft.ML.Data\Transforms\ColumnConcatenatingTransformer.cs:line 690
   at Microsoft.ML.Runtime.Data.DataViewUtils.Splitter.InPipe.Impl`1.Fill() in E:\A\_work\423\s\src\Microsoft.ML.Data\Data\DataViewUtils.cs:line 730
   at Microsoft.ML.Runtime.Data.DataViewUtils.Splitter.Consolidator.<>c__DisplayClass4_1.<ConsolidateCore>b__2() in E:\A\_work\423\s\src\Microsoft.ML.Data\Data\DataViewUtils.cs:line 417
```

Trace:
```
 	Microsoft.ML.Data.dll!Microsoft.ML.Runtime.Data.DataViewUtils.Splitter.Batch.SetAll(Microsoft.ML.Runtime.Data.DataViewUtils.Splitter.OutPipe[] pipes) Line 839	C#
 	Microsoft.ML.Data.dll!Microsoft.ML.Runtime.Data.DataViewUtils.Splitter.Cursor.MoveNextCore() Line 1114	C#
 	Microsoft.ML.Core.dll!Microsoft.ML.Runtime.Data.RootCursorBase.MoveNext() Line 70	C#
 	Microsoft.ML.Data.dll!Microsoft.ML.Transforms.Normalizers.NormalizingTransformer.Train(Microsoft.ML.Runtime.IHostEnvironment env, Microsoft.ML.Runtime.Data.IDataView data, Microsoft.ML.Transforms.Normalizers.NormalizingEstimator.ColumnBase[] columns) Line 375	C#
 	Microsoft.ML.Data.dll!Microsoft.ML.Runtime.Data.EstimatorChain<Microsoft.ML.Core.Data.ITransformer>.Fit(Microsoft.ML.Runtime.Data.IDataView input) Line 68	C#
 	Microsoft.ML.Data.dll!Microsoft.ML.Runtime.Data.EstimatorChain<Microsoft.ML.Core.Data.ITransformer>.Fit(Microsoft.ML.Runtime.Data.IDataView input) Line 68	C#
 	Microsoft.ML.Data.dll!Microsoft.ML.StaticPipe.Estimator<(Microsoft.ML.StaticPipe.Vector<float>, Microsoft.ML.StaticPipe.Key<uint, string>), (Microsoft.ML.StaticPipe.Scalar<string>, Microsoft.ML.StaticPipe.Scalar<string>, Microsoft.ML.StaticPipe.Vector<float>), Microsoft.ML.Core.Data.ITransformer>.Fit(Microsoft.ML.StaticPipe.DataView<(Microsoft.ML.StaticPipe.Vector<float>, Microsoft.ML.StaticPipe.Key<uint, string>)> view) Line 35	C#
[...]
```

### LINKED ISSUE #2678: Unintuitive error when giving array to OneHotEncoder
The following will throw a very ugly exception only when calling `Preview()` and not before:

```csharp
    public class MlDataEx
    {
      public string[] A { get; set; }
      public MlDataEx(string a)
      {
        this.A = new[] { a };
      }
    }

    public static void MlStuff()
    {
      var ctx = new MLContext();
      var dv = ctx.Data.ReadFromEnumerable(new[] { new MlDataEx("A") });
      var data = ctx.Transforms.Categorical.OneHotEncoding("A").Fit(dv).Transform(dv).Preview();
    }
```

### LINKED ISSUE #3757: KeyToVectorMappingTransformer: Index was outside the bounds of the array.
### System information

- **OS version/distro**: 1.0.0
- **.NET Version (eg., dotnet --info)**: 4.6.2

### Issue

- **What did you do?** Apply Normalization superviseBin and OneHot, 
- **What happened?** Got "Index was outside the bounds of the array."
- **What did you expect?** No error

### Source code / logs

Please paste or attach the code or logs or traces that would be helpful to diagnose the issue you are reporting.

```
            MLContext mlContext = new MLContext(seed: 0);
            var dataPath = "test1.csv";
            var featureName = "Features";
            var loader = mlContext.Data.CreateTextLoader(new[] 
            {
                new TextLoader.Column("int1", DataKind.Int64, 0),
                new TextLoader.Column("int2", DataKind.Int64, 1),
                new TextLoader.Column("Label", DataKind.Boolean, 2),
            }, hasHeader: true, separatorChar: ',');

            var data = loader.Load(dataPath);
            var learningPipeline = mlContext.Transforms.Conversion.ConvertType("int1", outputKind: DataKind.Single)
                    .Append(mlContext.Transforms.Conversion.ConvertType("int2", outputKind: DataKind.Single))
                    .Append(mlContext.Transforms.Concatenate(featureName, new string[] { "int1", "int2" }))
                    .Append(mlContext.Transforms.NormalizeSupervisedBinning(featureName, fixZero: false, maximumBinCount: 5, labelColumnName: "Label"))
                    .Append(mlContext.Transforms.Categorical.OneHotEncoding(featureName, outputKind: OneHotEncodingEstimator.OutputKind.Indicator));
            learningPipeline.Fit(data).Transform(data).Preview();
```

```
int1, int2, label
301, 2000, true
450, 3000, true
-300, 4000, true
300, 2000, false
115, 2000, false
115, 2000, false
```

I think it is related to [issue 1751](https://github.com/dotnet/machinelearning/issues/1751) And based on the discussion for this issue,  I tried 
- adding MapKeyToValue() but no help.
- OneHotEncode cannot be easily removed, we want to treat binning as categorical feature. 

```
            MLContext mlContext = new MLContext(seed: 0);
            var dataPath = "test1.csv";
            var featureName = "Features";
            var loader = mlContext.Data.CreateTextLoader(new[] 
            {
                new TextLoader.Column("int1", DataKind.Int64, 0),
                new TextLoader.Column("int2", DataKind.Int64, 1),
                new TextLoader.Column("Label", DataKind.Boolean, 2),
            }, hasHeader: true, separatorChar: ',');

            var data = loader.Load(dataPath);
            var learningPipeline = mlContext.Transforms.Conversion.ConvertType("int1", outputKind: DataKind.Single)
                    .Append(mlContext.Transforms.Conversion.ConvertType("int2", outputKind: DataKind.Single))
                    .Append(mlContext.Transforms.Conversion.MapValueToKey("Label"))
                    .Append(mlContext.Transforms.Concatenate(featureName, new string[] { "int1", "int2" }))
                    .Append(mlContext.Transforms.NormalizeSupervisedBinning(featureName, fixZero: false, maximumBinCount: 5, labelColumnName: "Label"))
                    .Append(mlContext.Transforms.Categorical.OneHotEncoding(featureName, outputKind: OneHotEncodingEstimator.OutputKind.Indicator))
                    .Append(mlContext.Transforms.Conversion.MapKeyToValue("Label"));
            learningPipeline.Fit(data).Transform(data).Preview();
```

### PR DESCRIPTION
**Pull request by codemzs**
fixes #3757
fixes #1751
fixes #2678

### CODE REVIEW

File: src/Microsoft.ML.Data/Transforms/KeyToVector.cs [hunk 1]
```diff
@@ -573,7 +573,14 @@ private ValueGetter<VBuffer<float>> MakeGetterInd(DataViewRow input, int iinfo)
                                 if (key >= (uint)size)
                                     continue;
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/Transforms/KeyToVector.cs]: It seems that we need to use `Indices` anyway, in case this condition is ever true, don't we?

Comment (yaeldekel) [reply src/Microsoft.ML.Data/Transforms/KeyToVector.cs]: @codemzs , it seems @Ivanidzo4ka had a PR for this: #2681 that hasn't been checked in for some reason.

File: test/Microsoft.ML.Tests/Transformers/CategoricalTests.cs [hunk 2]
```diff
@@ -101,6 +108,30 @@ public void CategoricalOneHotEncoding()
...
+                new TestClassWithLabel() { A = 450, B = 3000, Label = true },
+                new TestClassWithLabel() { A = -300, B = 4000, Label = true },
+                new TestClassWithLabel() { A = 300, B = 2000, Label = false },
+                new TestClassWithLabel() { A = 115, B = 2000, Label = false },
+                new TestClassWithLabel() { A = 115, B = 2000, Label = false }};
+
+            var mlContext = new MLContext();
+            var dataView = mlContext.Data.LoadFromEnumerable(data);
+            var pipe = mlContext.Transforms.Conversion.ConvertType("A", outputKind: DataKind.Single)
+        .Append(mlContext.Transforms.Conversion.ConvertType("B", outputKind: DataKind.Single))
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Transformers/CategoricalTests.cs]: (nit) the formatting here is off.  The `.Append` should be to the right of the line above it, for example:

```C#
            var pipe = mlContext.Transforms.Conversion.ConvertType("A", outputKind: DataKind.Single)
                .Append(mlContext.Transforms.Conversion.ConvertType("B", outputKind: DataKind.Single))
                .Append(mlContext.Transforms.Concatenate("Features", new string[] { "A", "B" }))
```

Comment (codemzs) [reply test/Microsoft.ML.Tests/Transformers/CategoricalTests.cs]: DOne.

**Pull request merged**